sirepo/pkcli/service.py

```diff
@@ -52,25 +59,159 @@ def flower():
+        except subprocess.CalledProcessError:
+            pkdlog('docker image {} not installed. Pulling...')
+            assert subprocess.call(('docker', 'image', 'pull', i)) == 0, \
+                'docker image pull failed'
 
-    with pkio.save_chdir(_run_dir()):
-        use_reloader = pkconfig.channel_in('dev')
-        app = server.init(use_reloader=use_reloader)
-        # avoid WARNING: Do not use the development server in a production environment.
-        app.env = 'development'
-        app.run(
-            host=cfg.ip,
-            port=cfg.port,
-            threaded=True,
-            use_reloader=use_reloader,
+        env.update(
+            SIREPO_JOB_DRIVER_MODULES='docker',
         )
 
+    def _env_local(env):
+        env.update(
+            SIREPO_JOB_DRIVER_MODULES='local',
+        )
+        assert 'SIREPO_JOB_DRIVER_MODULES' in env
+
+    def _env_nersc(env):
+        # TODO(e-carlin): Shouldn't nersc_user always be nagler?
+        assert nersc_proxy and nersc_user, \
+            'You need to supply a nersc_proxy and a nserc_user'
+        env.update(
+            SIREPO_JOB_DRIVER_MODULES='local:sbatch',
+            SIREPO_JOB_DRIVER_SBATCH_HOST='cori.nersc.gov',
+            SIREPO_JOB_DRIVER_SBATCH_SHIFTER_IMAGE='radiasoft/sirepo:sbatch',
+            SIREPO_JOB_DRIVER_SBATCH_SIREPO_CMD='/global/homes/{}/{}/.pyenv/versions/py3/bin/sirepo'.format(
+                nersc_user[0],
+                nersc_user,
+            ),
+            SIREPO_JOB_DRIVER_SBATCH_SRDB_ROOT='/global/cscratch1/sd/{sbatch_user}/sirepo-dev',
+            SIREPO_JOB_SUPERVISOR_SBATCH_POLL_SECS='15',
+            SIREPO_JOB_DRIVER_SBATCH_SUPERVISOR_URI='http://{}:8001'.format(nersc_proxy),
+            SIREPO_PKCLI_JOB_SUPERVISOR_IP='0.0.0.0',
+            SIREPO_SIMULATION_DB_SBATCH_DISPLAY='Cori@NERSC',
+        )
+
+    def _env_sbatch(env):
+        h = socket.gethostname()
+        env.update(
+            SIREPO_JOB_DRIVER_MODULES='local:sbatch',
+            SIREPO_JOB_DRIVER_SBATCH_HOST='{}'.format(
+                sbatch_host or h,
+            ),
+            SIREPO_JOB_DRIVER_SBATCH_SUPERVISOR_URI='http://{}:8001'.format(h),
+            SIREPO_PKCLI_JOB_SUPERVISOR_IP='0.0.0.0',
+            SIREPO_JOB_DRIVER_SBATCH_CORES='2',
+            SIREPO_JOB_DRIVER_SBATCH_SIREPO_CMD='{}/.pyenv/versions/py3/bin/sirepo'.format(os.getenv('HOME')),
+            SIREPO_JOB_DRIVER_SBATCH_SRDB_ROOT='/var/tmp/{sbatch_user}/sirepo',
+            SIREPO_JOB_SUPERVISOR_SBATCH_POLL_SECS='5',
+            SIREPO_SIMULATION_DB_SBATCH_DISPLAY='Vagrant Cluster',
+        )
+        if env['SIREPO_JOB_DRIVER_SBATCH_HOST'] == h:
+            # TODO(e-carlin): py3 has shutil.which()
+            assert distutils.spawn.find_executable('sbatch') is not None, \
+                'slurm not installed:  You need to run `radia_run slurm-dev`'
+        else:
+            env.update(SIREPO_JOB_DRIVER_SBATCH_CORES='4')
+
+    def _exit(*args):
```

Comment:
@robnagler  Once again I'm stumped by process management. This works in all cases (to the best of my knowledge) except one. When the "parent" flask process dies unexpectedly (kill -9). Then "reloader" flask process process becomes a zombie. I'm unsure how to solve this. I'm using `preexc_fn=os.setsid` which I thought creates a process group so when I kill the parent then the child should die but that isn't the case. Before I dig into how flask might be monkeying around with the processes I figured I'd I ask you if you see a better way to solve this.

Reply:
I think there's too much work. Just use kill and don't set a session (noted elsewhere). 

The only time this doesn't work is sometimes with MPI, but let's not worry about that in this case. This code is either going to be used by a developer or inside a Docker container (curl installer) iwc the user will stop sirepo with docker rm -f.

Reply:
Ugh. Much simpler following what you described. Thanks for the help.